pkg/virt-launcher/virtwrap/nichotplug.go

```diff
@@ -87,9 +89,44 @@ func (vim *virtIOInterfaceManager) hotplugVirtioInterface(vmi *v1.VirtualMachine
 	return nil
 }
 
-func domainInterfaceFromNetwork(domain *api.Domain, network v1.Network) *api.Interface {
-	for _, iface := range domain.Spec.Devices.Interfaces {
-		if iface.Alias.GetName() == network.Name {
+func (vim *virtIOInterfaceManager) hotUnplugVirtioInterface(vmi *v1.VirtualMachineInstance, currentDomain *api.Domain) error {
+	for _, domainIface := range interfacesToHotUnplug(vmi.Spec.Domain.Devices.Interfaces, currentDomain.Spec.Devices.Interfaces) {
+		log.Log.Infof("preparing to hot-unplug %s", domainIface.Alias.GetName())
+
+		ifaceXML, err := xml.Marshal(domainIface)
+		if err != nil {
+			return err
+		}
+
+		if derr := vim.dom.DetachDeviceFlags(strings.ToLower(string(ifaceXML)), affectLiveAndConfigLibvirtFlags); derr != nil {
+			log.Log.Reason(derr).Errorf("libvirt failed to detach interface %s: %v", domainIface.Alias.GetName(), derr)
+			return derr
+		}
+	}
+	return nil
+}
+
+func interfacesToHotUnplug(vmiSpecInterfaces []v1.Interface, domainSpecInterfaces []api.Interface) []api.Interface {
+	ifaces2remove := netvmispec.FilterInterfacesSpec(vmiSpecInterfaces, func(i v1.Interface) bool {
+		return i.State == v1.InterfaceStateAbsent
+	})
+	var domainIfacesToRemove []api.Interface
+	for _, vmiIface := range ifaces2remove {
+		if domainIface := lookupDomainInterfaceByName(domainSpecInterfaces, vmiIface.Name); domainIface != nil {
+			if domainIface.Target != nil &&
+				strings.HasPrefix(domainIface.Target.Device, "tap") &&
+				domainIface.Target.Device[3:] == namescheme.GenerateHashedInterfaceName(vmiIface.Name)[3:] {
```

Comment:
This is what I meant by using regex:
```
func hasDeviceWithHashedTapName(target *api.InterfaceTarget, vmiIface v1.Interface) bool {
        if target.Device != nil {
                if match, err := regexp.MatchString("^tap\d$", target.Device); err == nil {
                         return match                
                }
        }
	return false
}
```

The current implementation generates the tap device name, trims the generated name and actual name prefixes, and compares the names.
I think using regex is more efficient than generating the name and trimming the prefixes.

Reply:
> `"^tap\d$"`

This seems to match ordinal names with only one digit (i.e. up to 10 interfaces).
We could add another digit and then negate it, but it seems over complex to a common reader.
The fact that the prefix is not changed makes this less attractive IMO.

Reply:
Sorry for confusing you, what I had in mind is making this function to check if a tap device has an *ordinal* name (`hasDeviceWithOrdinalTapName`).
But this can still work as is, the return values should the opposite.

We can change the pattern to `^tap\d+$` (add "+") similar to how ordinal iface names are matched today (`^net\d+$`).
IMO it's easier to understand than the name generation and prefix trimming.

Reply:
> We can change the pattern to ^tap\d+$ (add "+") similar to how ordinal iface names are matched today (^net\d+$).

This will also be incorrect. `^tap\d+$` matches also hashed names.

Reply:
Well, then we can use `^tap\d{1,2}$` as you suggested, IMO it's still a better option.